// lib/views/LayoutView.js
'use strict';

var _ = require('underscore');
var View = require('lib/views/View');

module.exports = View.extend({
  name: 'LayoutView',

  constructor: function(options) {
    options = options || {};
    this.views = options.views || {};
    View.apply(this, arguments);
    this.keepEl = true;
  },

  createSubView: function(view) {
    //console.log('create', view)
    //if (!view.instance) {
    view.instance = new view.constructor(view.options);
    //}
  },

  renderSubViews: function() {
    _.each(this.views, this.renderSubView, this);
  },

  renderSubView: function(view, el) {
    this.createSubView(view);
    view.instance.setElement(this.$el.find(el)).render();
  },

  removeSubViews: function() {
    _.each(this.views, this.removeSubView, this);
  },

  removeSubView: function(view) {
    if (view.instance) {
      view.instance.remove();
    }
  },

  render: function() {
    View.prototype.render.call(this);
    this.renderSubViews();
    return this;
  },

  remove: function() {
    this.removeSubViews();
    View.prototype.remove.call(this);
    return this;
  }

});
